File: 04_Array_24.py
def flip(arr, i):
    start = 0
    while start < i:
        temp = arr[start]
        arr[start] = arr[i]
        arr[i] = temp
        start += 1
        i -= 1


def findMax(arr, n):
    mi = 0
    for i in range(0, n):
        if arr[i] > arr[mi]:
            mi = i
    return mi


def pancakeSort(arr, n):
    curr_size = n
    while curr_size > 1:

        mi = findMax(arr, curr_size)

        if mi != curr_size - 1:

            flip(arr, mi)
            if mi != 0:
                printArray(arr, n)

            flip(arr, curr_size - 1)
            printArray(arr, n)
        curr_size -= 1


def printArray(arr, n):
    for i in range(0, n):
        print("%d" % (arr[i]), end=" ")
    print('')


n = int(input())
arr = [int(e) for e in input().split()]
printArray(arr, n)
pancakeSort(arr, n)

File: solution/04_Array_24.cpp
#include <iostream>
using namespace std;

int flip(int arr[], int i)
{
    int start = 0;
    while (start < i)
    {
        int temp = arr[start];
        arr[start] = arr[i];
        arr[i] = temp;
        start += 1;
        i -= 1;
    }
    return 0;
}

int findMax(int arr[], int n)
{
    int mi = 0;
    for (int i = 0; i < n; i++)
    {
        if (arr[i] > arr[mi])
        {
            mi = i;
        }
    }
    return mi;
}

int printArray(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        if (i == n - 1)
        {
            cout << arr[i];
        }
        else
        {
            cout << arr[i] << ' ';
        }
    }
    cout << endl;
    return 0;
}

int pancakeSort(int arr[], int n)
{
    int curr_size = n;
    while (curr_size > 1)
    {
        int mi = findMax(arr, curr_size);

        if (mi != curr_size - 1)
        {
            flip(arr, mi);
            if (mi != 0)
            {
                printArray(arr, n);
            }
            flip(arr, curr_size - 1);
            printArray(arr, n);
            mi = curr_size - 1;
        }
        curr_size -= 1;
    }
    return 0;
}

int main()
{
    int n;
    cin >> n;

    int arr[n];
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }

    printArray(arr, n);
    pancakeSort(arr, n);
    return 0;
}